### PR TITLE
generator.php

### PR DESCRIPTION
**Pull request by AndreiLupuleasa**
- overload classes method added
templates
	- request overload created
	- response overload created

### CODE REVIEW

**Comment by dcarbone** [comment]
@AndreiLupuleasa: So while I like the initiative, I'd like for you to consider a bit more how this would actually be used.  There is not much point in creating templates for the overloading classes as it is assumed the classes have already been created by the implementer.  We are mostly interested in being able to return to the user, rather than a direct instance of our generated class, an instance of the class they specify.

Think of it like this:

I want to have a helper method on `ListZonesResponse` that returns a value from a specific tag.  I would create a simple class like this:

```php
class OverloadedListZonesResponse extends CloudStackResponse\ListZonesResponse {
    public function displayName(): string {
        // ... logic to pull from tags
    }
}
```

...and that would be the end of it.  The rest of the class should be created and built as per the normal generated output.

**Comment by dcarbone** [comment]
What would be the purpose of listing the methods that are overloaded?  The generator doesn't care.

> I don't think I understand this (Shouldn't the extended class contain only the overloaded methods?)

Yes, that is my exact point.

**Comment by AndreiLupuleasa** [comment]
Please excuse my stupid question @dcarbone.
I've modified my code and made a new pull request.

**Comment by AndreiLupuleasa** [comment]
It wasn't necessary to make another pull request, was it?

**Comment by AndreiLupuleasa** [comment]
If I make another commit the pull request has the new commit.

**Comment by dcarbone** [comment]
apologies for the delay, forgot to actually submit the review :)

File: src/Generator.php [hunk 1]
```diff
@@ -479,6 +481,38 @@ protected function writeOutRequestModels()
...
+                $this->writeFile(
+                    $path . '/' . $overloadedClass->getOverload() . '.php',
+                    $template->render([
+                        'obj'       => $this->sharedObjectMap[strtolower($overloadedClass->getName())],
+                        'overloadClassName' => $overloadedClass->getOverload(),
+                    ])
+                );
+            }
+		}
+    }
```
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply src/Generator.php]: This is how my code looks in sublime
image

[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply src/Generator.php]: image

[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply src/Generator.php]: image

File: templates/models/response.php.twig [hunk 2]
```diff
@@ -26,7 +26,7 @@
  * Class {{ obj.className }}
  * @package {{ namespace_path('CloudStackResponse', true) }}
 {% if api is defined %}{{ since_tag_line(api.since, 0, true) }}{% endif %} */
-class {{ obj.className }}{% if obj.shared == false %} implements {% if isPageableResponse %}CloudStackPageableListResponse{% elseif isListResponse %}CloudStackListResponse{% else %}CloudStackResponseInterface{% endif %}{% endif %}
+class {% if overloadClassName is defined %}{{overloadClassName}} extends {% endif %}{{ obj.className }}{% if obj.shared == false %} implements {% if isPageableResponse %}CloudStackPageableListResponse{% elseif isListResponse %}CloudStackListResponse{% else %}CloudStackResponseInterface{% endif %}{% endif %}
 
 {
 {% for prop in obj.properties %}
```
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply templates/models/response.php.twig]: I didn't removed all those lines in red
image

[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply templates/models/response.php.twig]: I didn't removed anything, I just added an if statement.

File: src/Generator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply src/Generator.php]: I've checked ident using spaces and it looks the same.
https://superuser.com/questions/905820/why-do-my-tabs-in-sublime-text-appear-so-far-off-in-github

File: templates/models/request.php.twig [hunk 3]
```diff
@@ -45,7 +45,7 @@ use {{ namespace_path('CloudStackHelpers') }};{% endif %}
  * Class {{ api.requestClassname }}
  * @package {{ namespace_path('CloudStackRequest', true) }}
 {{ since_tag_line(api.since, 0, true) }} */
```
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply templates/models/request.php.twig]: I didn't erased the line

File: templates/models/response.php.twig (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply templates/models/response.php.twig]: image

[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply templates/models/response.php.twig]: image

File: templates/models/request.php.twig (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply templates/models/request.php.twig]: image

[the remark this answers is not in the document]
Comment (dcarbone) [reply templates/models/request.php.twig]: Yes apologies, the red markup is due to Github's syntax highlighting going nuts.

File: src/Generator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (AndreiLupuleasa) [reply src/Generator.php]: It should be ok now

**Comment by AndreiLupuleasa** [comment]
@dcarbone  I made the requested changes in request.php on line 181 and in generator.php

**Comment by dcarbone** [comment]
Great!  Have you run the generator to test if it works?

**Comment by AndreiLupuleasa** [comment]
If the class is not overloaded than the method looks like this:
```php
public function getCloudStackResponseModel(): string
{
    return '\\ENA\\CloudStack\\CloudStackResponse\\ListVirtualMachinesResponse';
}
```
Should my method look like this?
```php
public function getCloudStackResponseModel(): string
{
    return '\\MyNamespace\\DoSomethingCrazyRequest';
}
```

**Pull request merged**